pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -571,6 +579,65 @@ func (a Adapter) generateDeploymentObjectMeta(deployment *appsv1.Deployment, lab
 	}
 }
 
+// deleteRemoteResourcesNotPresentInDevfile compares the list of Devfile K8s component and remote K8s resources
+// and removes the remote resources not present in the Devfile;
+// it ignores the core components (such as deployments, svc, pods; all resources with `component:<something>` label)
+func (a Adapter) deleteRemoteResourcesNotPresentInDevfile(labels map[string]string) error {
+	allRemoteK8sResources, err := a.kubeClient.GetAllResourcesFromSelector(util.ConvertLabelsToSelector(labels), a.kubeClient.GetCurrentNamespace())
+	if err != nil {
+		return fmt.Errorf("unable to fetch remote kubernetes resources: %w", err)
+	}
+	var remoteK8sResources []unstructured.Unstructured
+	// Filter core components
+	for _, remoteK := range allRemoteK8sResources {
+		if !odolabels.IsCoreComponent(remoteK.GetLabels()) {
+			remoteK8sResources = append(remoteK8sResources, remoteK)
+		}
+	}
+	devfileK8sResources, err := devfile.GetKubernetesComponentsToPush(a.Devfile)
+	if err != nil {
+		return fmt.Errorf("unable to obtain devfile kubernetes resources: %w", err)
+	}
+
+	// convert all devfileK8sResources to unstructured data
+	var devfileK8sResourcesUnstructured []unstructured.Unstructured
+	for _, devfileK := range devfileK8sResources {
+		devfileKUnstructured, err := libdevfile.GetK8sComponentAsUnstructured(a.Devfile, devfileK.Name, a.Context, devfilefs.DefaultFs{})
+		if err != nil {
+			return fmt.Errorf("unable to obtain unstructured data: %w", err)
+		}
+		devfileK8sResourcesUnstructured = append(devfileK8sResourcesUnstructured, devfileKUnstructured)
+	}
+
+	var objectsToRemove []unstructured.Unstructured
+	for _, remoteK := range remoteK8sResources {
+		matchFound := false
+		for _, devfileK := range devfileK8sResourcesUnstructured {
+			if devfileK.GetKind() == remoteK.GetKind() && devfileK.GetName() == remoteK.GetName() && remoteK.GetNamespace() == a.kubeClient.GetCurrentNamespace() {
```

Comment:
You will also need to test the group, or you could mix resources of same kind in different groups.